numpy/core/src/multiarray/ctors.c

```diff
@@ -1232,12 +1246,12 @@ PyArray_NewLikeArray(PyArrayObject *prototype, NPY_ORDER order,
             break;
     }
 
-    /* If it's not KEEPORDER, this is simple */
-    if (order != NPY_KEEPORDER) {
+    /* If it's not KEEPORDER, or there is a shape change, this is simple */
+    if (order != NPY_KEEPORDER || (new_shape && ndim != PyArray_NDIM(prototype))) {
```

Comment:
You can simplify this back to what it was before now that `NPY_KEEPORDER` implies no shape change, right?